### PR TITLE
Add an option to disable write stall

### PR DESCRIPTION
**Pull request by Connor1996**
- Add an option `disable-write-stall` to disable write stall. It avoids RocksDB from being the write stall state, but still keeps the logic of decreasing/increasing compaction threads unchanged as before.
- Use `base_background_compactions` to decide `max_compactions` when decrease/increase compaction threads.

### CODE REVIEW

File: db/db_impl/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -1986,7 +1988,7 @@ DBImpl::BGJobLimits DBImpl::GetBGJobLimits(int max_background_flushes,
   }
   if (!parallelize_compactions) {
     // throttle background compactions until we deem necessary
-    res.max_compactions = 1;
+    res.max_compactions = std::max(1, base_background_compactions);
```
Comment (tabokie) [inline db/db_impl/db_impl_compaction_flush.cc]:
```suggestion
    res.max_compactions = std::max(1, std::min(base_background_compactions, res.max_compactions));
```

**Review by tabokie — APPROVED** [review]
Rest looks good.

File: options/cf_options.cc [hunk 2]
```diff
@@ -148,6 +148,8 @@ void MutableCFOptions::Dump(Logger* log) const {
       prefix_extractor == nullptr ? "nullptr" : prefix_extractor->Name());
   ROCKS_LOG_INFO(log, "                 disable_auto_compactions: %d",
                  disable_auto_compactions);
+  ROCKS_LOG_INFO(log, "                 disable_write_stall: %d",
+                 disable_write_stall);
```
Comment (yiwu-arbug) [inline options/cf_options.cc:152]: should print `base_background_compactions` as well

**Review by yiwu-arbug — APPROVED** [review]
LGTM overall.

Have you tested both of `disable_write_stall` and `base_background_compactions` are dynamic changeable?

**Comment by Connor1996** [comment]
> LGTM overall.
> 
> Have you tested both of `disable_write_stall` and `base_background_compactions` are dynamic changeable?

Add two test cases to cover the two configs.

File: db/column_family.cc [hunk 3]
```diff
@@ -829,7 +829,7 @@ WriteStallCondition ColumnFamilyData::RecalculateWriteStallConditions(
           name_.c_str(), vstorage->estimated_compaction_needed_bytes(),
           write_controller->delayed_write_rate());
     } else {
-      assert(write_stall_condition == WriteStallCondition::kNormal);
+      assert(write_stall_condition == WriteStallCondition::kNormal || mutable_cf_options.disable_write_stall);
```
Comment (tabokie) [inline db/column_family.cc:832]: @Connor1996 Minor: should be `else if (write_stall_condition == WriteStallCondition::kNormal)` ?

Comment (Connor1996) [reply db/column_family.cc:832]: What's the difference

Comment (tabokie) [reply db/column_family.cc:832]: Each branch has its own precondition, which is "`write_stall_condition` is unset" here. For this specific spot, breaking the precondition might not cause any trouble, but it's a mistake nonetheless.

Comment (Connor1996) [reply db/column_family.cc:832]: `disable_write_stall` is a kind of "write_stall_condition is unset", using `else if` needs to duplicate the logic in `else` branch.

**Pull request merged**